packages/braintree-integration/src/braintree-fastlane/braintree-fastlane-utils.ts

```diff
@@ -0,0 +1,541 @@
+            const cart = state.getCartOrThrow();
+            const customer = state.getCustomer();
+            const billingAddress = state.getBillingAddress();
+
+            const customerEmail = email || customer?.email || billingAddress?.email || '';
+
+            const { customerContextId } = await lookupCustomerByEmail(customerEmail);
+
+            if (!customerContextId) {
+                // Info: we should clean up previous experience with default data and related authenticationState
+                await this.paymentIntegrationService.updatePaymentProviderCustomer({
+                    authenticationState: BraintreeConnectAuthenticationState.UNRECOGNIZED,
+                    addresses: [],
+                    instruments: [],
+                });
+
+                this.browserStorage.setItem('sessionId', cart.id);
+
+                return;
+            }
+
+            const { authenticationState, profileData } = await triggerAuthenticationFlow(
+                customerContextId,
+            );
+
+            if (authenticationState === BraintreeConnectAuthenticationState.CANCELED) {
+                await this.paymentIntegrationService.updatePaymentProviderCustomer({
+                    authenticationState,
+                    addresses: [],
+                    instruments: [],
+                });
+
+                this.browserStorage.removeItem('sessionId');
+
+                return;
+            }
+
+            const shippingAddresses =
+                this.mapPayPalToBcAddress(profileData.addresses, profileData.phones) || [];
+            const paypalBillingAddresses = this.getPayPalBillingAddresses(profileData);
+            const billingAddresses =
+                this.mapPayPalToBcAddress(paypalBillingAddresses, profileData.phones) || [];
+            const instruments = this.mapPayPalToBcInstrument(methodId, profileData.cards) || [];
+            const addresses = this.mergeShippingAndBillingAddresses(
+                shippingAddresses,
+                billingAddresses,
+            );
+
+            this.browserStorage.setItem('sessionId', cart.id);
+
+            await this.paymentIntegrationService.updatePaymentProviderCustomer({
+                authenticationState,
+                addresses,
+                instruments,
+            });
+
+            if (billingAddresses.length > 0) {
+                await this.paymentIntegrationService.updateBillingAddress(billingAddresses[0]);
+            }
+
+            if (shippingAddresses.length > 0 && cart.lineItems.physicalItems.length > 0) {
+                await this.paymentIntegrationService.updateShippingAddress(shippingAddresses[0]);
+            }
+        } catch (error) {
+            // TODO: we should figure out what to do here
+            // TODO: because we should not to stop the flow if the error occurs on paypal side
+        }
+    }
+
+    async runPayPalFastlaneAuthenticationFlowOrThrow(email?: string): Promise<void> {
```

Comment:
I don't really sure that it is a good idea to copy&paste prev method where only 2 lines of code were changed. Lets create a single method (runPayPalAuthenticationFlowOrThrow) and decide what Fastlane or Connect to use there

Reply:
They are not actually the same, they used different methods because Fastlane have different interfaces

Reply:
I'll refactor it anyway